doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+ */
+ELEKTRA_CONST ElektraKeyset * elektraKeyGetMeta (const ElektraKey * key);
+
+ElektraErrorCode elektraKeyIsBelow (const ElektraKey * first, const ElektraKey * second);
+ElektraErrorCode elektraKeyCompare (const ElektraKey * first, const ElektraKey * second);
+
+ElektraKey * elektraKeyCopy (ElektraKey * dest, const ElektraKey * src, ElektraKeyCopyFlag flags);
+static inline ElektraKey * elektraKeyDup (ElektraKey * key, ElektraKeyCopyFlag flags)
+{
+    ElektraKey * dest = flags & ELEKTRA_KEY_CP_NAME
+        ? elektraKeyNew (elektraKeyGetName (key))
+        : elektraKeyNew (&(ElektraKeyname){.ns = ELEKTRA_NS_CASCADING, .name = "", .size = 0});
+    return elektraKeyCopy (dest, key, (ElektraKeyCopyFlag)(flags & ~ELEKTRA_KEY_CP_NAME));
+}
+
+
+ElektraKeyset * elektraKeysetNew (size_t prealloc);
+ElektraKeyset * elektraKeysetRetain (const ElektraKeyset * ks);
+void elektraKeysetRelease (const ElektraKeyset * ks);
+
+size_t elektraKeysetSize (ElektraKeyset * ks);
+
+ElektraErrorCode elektraKeysetInsert (ElektraKeyset * ks, ElektraKey * key);
+ElektraErrorCode elektraKeysetInsertAll (ElektraKeyset * ks, ElektraKeyset * other);
+static inline ElektraErrorCode elektraKeysetInsertAndRelease (ElektraKeyset * ks, ElektraKey * key)
+{
+    ElektraErrorCode error = elektraKeysetInsert (ks, key);
+    elektraKeyRelease (key);
+    return error;
+}
+
+ElektraKey * elektraKeysetGet (ElektraKeyset * ks, size_t index);
+ElektraKeyset * elektraKeysetGetRange (ElektraKeyset * ks, size_t start, size_t end);
+
+void elektraKeysetRemove (ElektraKeyset * ks, size_t index);
+void elektraKeysetRemoveRange (ElektraKeyset * ks, size_t start, size_t end);
+static inline void elektraKeysetClear (ElektraKeyset * ks)
+{
+    elektraRemoveRange (ks, 0, elektraKeysetSize (ks));
+}
+
+size_t elektraKeysetLookup (ElektraKeyset * ks, const ElektraKeyname * name);
+
+size_t elektraKeysetFindHierarchy (ElektraKeyset * ks, const ElektraKeyname * root, size_t * end);
+```
+
+### Private headers
+
+```c
+struct ElektraKeynameCow {
+    struct ElektraKeyname name;
+
+    uint16_t refs;
+
+    bool shouldFree : 1;
+    int reserved : 15;
+};
+
+struct ElektraKeyvalueCow {
+    struct ElektraKeyvalue value;
+
+    uint16_t refs;
+
+    bool shouldFree : 1;
+    int reserved : 15;
+};
+
+struct ElektraKeyset;
+
+struct ElektraKey {
+    struct ElektraKeynameCow * name;
+    struct ElektraKeyvalueCow * value;
+    struct ElektraKeyset * meta;
+
+    uint16_t refs;
+    uint16_t nameLock;
+
+    bool shouldFree : 1;
+    bool needsSync : 1;
+    int reserved : 14;
+};
+
+struct ElektraKeyset {
+    struct ElektraKeysetDataCow * data;
+
+    uint16_t refs;
+
+    bool shouldFree : 1; // used for mmap
+    int reserved : 15;
+};
+```
+
+### Usage Examples
+
+```c
+// Create key (Note: better ways below via additional APIs)
+ElektraKey * k1 = elektraKeyNew (&(ElektraKeyname){
+    .ns = ELEKTRA_NS_SYSTEM,
+    .name = "foo\0bar\0baz",
+    .size = 14
+});
+
```

Comment:
Looks terrible.

Reply:
That's a very personal opinion. I actually find it much more pleasing to look at than the current var-arg stuff in `keyNew`. Not least, because this has a very clear way of formatting it properly.

Reply:
I thought also rough and unspecific feedback is welcomed. I don't understand the reply, where exactly should I look?

Reply:
> I thought also rough and unspecific feedback is welcomed

Yes, "looks terrible" is not feedback at all. I can't make use of this in anyway. I don't know which part you don't like, or what you'd like better.

> I don't understand the reply, where exactly should I look?

I don't understand the question...

What I personally said was, I find this `elektraKeyNew` call more less terrible (i.e. "more pleasing to look at") than the current `keyNew (..., KEY_VALUE, ... KEY_META, ...)` stuff.